src/components/PDFView/index.js

```diff
@@ -49,21 +67,68 @@ class PDFView extends Component {
     }
 
     /**
-     * Upon successful document load, set the number of pages on PDF,
+     * Upon successful document load, combine an array of page viewports,
+     * set the number of pages on PDF,
      * hide/reset PDF password form, and notify parent component that
      * user input is no longer required.
      *
-     * @param {*} {numPages} No of pages in the rendered PDF
+     * @param {*} pdf The PDF file instance
      * @memberof PDFView
      */
-    onDocumentLoadSuccess({numPages}) {
-        this.setState({
-            numPages,
-            shouldRequestPassword: false,
-            isPasswordInvalid: false,
+    onDocumentLoadSuccess(pdf) {
+        const numPages = pdf.numPages;
+
+        Promise.all(
+            _.times(numPages, (index) => {
+                const pageNumber = index + 1;
+
+                return pdf.getPage(pageNumber).then((page) => page.getViewport({scale: 1}));
+            }),
+        ).then((pageViewports) => {
+            this.setState({
+                pageViewports,
+                numPages,
+                shouldRequestPassword: false,
+                isPasswordInvalid: false,
+            });
         });
     }
 
+    /**
+     * Calculates a proper page height. The method should be called only when there are page viewports.
+     * It is based on a ratio between the specific page viewport width and provided page width.
+     * Also, the app should take into account the page borders.
+     * @param {*} pageIndex
+     * @returns {Number}
+     */
+    calculatePageHeight(pageIndex) {
+        if (this.state.pageViewports.length === 0) {
+            Log.warn('Dev error: calculatePageHeight() in PDFView called too early');
+
+            return 0;
+        }
+
+        const pageViewport = this.state.pageViewports[pageIndex];
+        const pageWidth = this.calculatePageWidth();
+        const scale = pageWidth / pageViewport.width;
+        const actualHeight = pageViewport.height * scale + PAGE_BORDER * 2;
```

Comment:
Looks like we're using the [virtual list](https://www.patterns.dev/posts/virtual-lists) concept here. Just out of curiosity, what's stopping us from using this solution on native?

Reply:
AFAIK:

> Looks like we're using the [virtual list](https://www.patterns.dev/posts/virtual-lists) concept here. Just out of curiosity, what's stopping us from using this solution on native?

The libraries web and native are using are completely different.
As this issue happens only on mSafari, it's fine to optimize only react-pdf (web).
react-native-pdf (native) is "perfect" library and we haven't found any need of customization yet.

> If I'm not wrong, react-window is a web only library. Why not use FlatList and render x pages only?

FlatList can only be used on web combined with react-native-web. As react-pdf is a pure web library (not depending on react-native-web), it doesn't make sense to integrate with FlatList.
Using FlatList means that we should customize full PDF view without the benefit of core features in react-pdf.

Reply:
In short, `FlatList` does not work the same way on web as it does on native platforms.